CoreRemoting/RemotingClient.cs

```diff
@@ -316,11 +316,11 @@ public async Task DisconnectAsync(bool quiet = false)
                     await _goodbyeCompletedTaskSource.Task.Expire(10).ConfigureAwait(false);
             }
 
-            // lock (_syncObject) // TODO: why we are locking here?
+            lock (_syncObject)
             {
                 var channel = _channel;
                 if (channel != null)
-                    await channel.DisconnectAsync().ConfigureAwait(false);
+                    channel.DisconnectAsync().ConfigureAwait(false);
```

Comment:
Because of the lock, we can't await here.
I would prefer not to leave the unawaited task here.
It's a possible memory leak, exception leak, etc.

Perhaps we can find a way to replace the lock with a task-based awaitable synchronization primitive?

Reply:
yes, after solving the problem with concurrent access to _tcpClient, await must be returned and lock removed

Reply:
My solution to the async/await/lock incompatibility: #143

Reply:
Hmm, looks like _syncObject is widely used in synchronous methods or RemotingClient.

Also, it seems to be used to protect many different things, i.e. _sessionId, _activeCalls and _channel.

Reply:
Granular locks: PR #149.